### DISCUSSION TITLE
How to migrate a seperate db file and ATTACH data.db to it?

### DISCUSSION THREAD

**Discussion by lumosminima** [pocketbase/pocketbase#5621] [Q&A] [answered]
Don't know how to word it exactly, but basically I want to allow a single tenant to export their data from the `data.db` which houses multiple tenants data. In #5549, I learned that I could pluck the data migration functions generated by migration command to scaffold the db structure. I haven't worked directly with dbs before so don't know what are the best practices on this type of workflow but after some research it seamed like doing ATTACH on data.db will be the simplest since I only need read-only access to copy the most current data of a single user, so I came up with following experimental handler with non_cgo db connect code which is invoked on a custom route `/v1/export`:
```go
package export

import (
	"fmt"
	"log"
	"os"
	"path/filepath"
	"time"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/daos"
	_ "modernc.org/sqlite"
)

func HandleExport(app core.App, c echo.Context) error {
	userId := c.QueryParam("userId")
	if userId == "" {
		return nil
	}
	log.Println("this is userId", userId)

	appDbPath := filepath.Join(app.DataDir(), "data.db")
	subsetDbPath := filepath.Join(app.DataDir(), fmt.Sprintf("%s-%s.db", "datatest", userId))

	 if err := os.Remove(subsetDbPath); err != nil && !os.IsNotExist(err) {
            log.Println("remove error", err)
            return err
         }
	newSubsetDb, err := connectDB(subsetDbPath)
	if err != nil {
		log.Println("connectdb error", err)
		return err
	}

	newSubsetDbDao, scaffoldErr := scaffoldDb(newSubsetDb)

	if scaffoldErr != nil {
		return scaffoldErr
	}

	_, attachErr := newSubsetDbDao.DB().NewQuery(`ATTACH DATABASE {:path} AS app_db`).Bind(dbx.Params{"path": appDbPath}).Execute()
	if attachErr != nil {
		log.Println("attach error", attachErr)
	}

	newSubsetDbDao.RunInTransaction(func(txDao *daos.Dao) error {
		queries := []string{
			`INSERT OR REPLACE INTO users (id, name, email, username, passwordHash, tokenKey)
             SELECT id, name, email, username, "", "" FROM app_db.users
             WHERE id = {:userId};`,

			`INSERT OR REPLACE INTO roles SELECT * FROM app_db.roles;`,

			`INSERT OR REPLACE INTO permissions SELECT * FROM app_db.permissions;`,

			`INSERT OR REPLACE INTO roles_permissions SELECT * FROM app_db.roles_permissions;`,

			`INSERT OR REPLACE INTO spaces_roles
             SELECT * FROM app_db.spaces_roles
             WHERE user_id = {:userId};`,

			`INSERT OR REPLACE INTO spaces (id, name, space_role_ids, created, updated)
             SELECT p.id, p.name, p.space_role_ids, p.created, p.updated
             FROM app_db.spaces p
             JOIN app_db.spaces_roles pr ON p.id = pr.space_id
             WHERE pr.user_id = {:userId};`,

            // ....

		for _, query := range queries {
			_, err := txDao.DB().NewQuery(query).Bind(dbx.Params{"userId": userId}).Execute()
			if err != nil {
				return log.Printf("failed to execute query: %s", err)
			}
		}

		return nil
	})
	_, detachErr := newSubsetDbDao.DB().NewQuery(`DETACH DATABASE app_db;`).Execute()
	if detachErr != nil {
		log.Printf("failed to detach database: %s", detachErr)
		return detachErr
	}

	_, vacErr := newSubsetDbDao.DB().NewQuery("VACUUM").Execute()
	if vacErr != nil {

		log.Println("vacuum subset error", vacErr)
	}

	newSubsetDb.Close()

	return c.File(subsetDbPath)
}

func connectDB(dbPath string) (*dbx.DB, error) {

	// Note: the busy_timeout pragma must be first because
	// the connection needs to be set to block on busy before WAL mode
	// is set in case it hasn't been already set by another connection.
	pragmas := "?_pragma=busy_timeout(10000)&_pragma=journal_mode(WAL)&_pragma=journal_size_limit(200000000)&_pragma=synchronous(NORMAL)&_pragma=foreign_keys(ON)&_pragma=temp_store(MEMORY)&_pragma=cache_size(-16000)"

	db, err := dbx.Open("sqlite", dbPath+pragmas)
	if err != nil {
		log.Println("connect error", err)
		return nil, err
	}

	return db, nil
}

func scaffoldDb(db *dbx.DB) (*daos.Dao, error) {

	_, err := db.NewQuery(`

			CREATE TABLE IF NOT EXISTS {{_collections}} (
				[[id]]         TEXT PRIMARY KEY NOT NULL,
				[[system]]     BOOLEAN DEFAULT FALSE NOT NULL,
				[[type]]       TEXT DEFAULT "base" NOT NULL,
				[[name]]       TEXT UNIQUE NOT NULL,
				[[schema]]     JSON DEFAULT "[]" NOT NULL,
				[[indexes]]    JSON DEFAULT "[]" NOT NULL,
				[[listRule]]   TEXT DEFAULT NULL,
				[[viewRule]]   TEXT DEFAULT NULL,
				[[createRule]] TEXT DEFAULT NULL,
				[[updateRule]] TEXT DEFAULT NULL,
				[[deleteRule]] TEXT DEFAULT NULL,
				[[options]]    JSON DEFAULT "{}" NOT NULL,
				[[created]]    TEXT DEFAULT (strftime('%Y-%m-%d %H:%M:%fZ')) NOT NULL,
				[[updated]]    TEXT DEFAULT (strftime('%Y-%m-%d %H:%M:%fZ')) NOT NULL
			);


		`).Execute()
	if err != nil {
		return nil, err
	}

	dao, dberr := NewScaffoldedDb(db) // migration function that does ImportCollection() and also returns the new dao
	if dberr != nil {
		return nil, dberr
	}
	return dao, nil
}

```
This created a new `datatest-<user-id>.db` in pb_data folder which contained a single users data with same collection structure and it worked but soon after I added a new record to some collection in data.db and invoked this endpoint again the updated record would sometimes be in the newly created db and other times not and in those times I would also get Disk I/O errors like file no found in debug console for data.db queries from admin console. Eventually the record sometimes persists and sometimes it only showed on the admin console collections and after restarting the app it disappears. So, is there some recommended and fool proof way to achieve this and also how can I do the above in `0.23.0-rc2` since now there is no `daos.New()` I tried creating a whole new app in the endpoint with custom DBConnect path and then app.Start() but it gives a port already binded error.

**Comment by ganigeorgiev** [comment]
You don't really use the `Dao` in the above because it seems that the export operations are through `Dao.DB()` which returns the dbx.Builder instance (aka. the result of your `connectDB` function) and you can use that directly.

For starting a transaction with the `dbx` builder you can check https://github.com/pocketbase/dbx#using-transactions.

In general having slow mixed (read&write) transactions is not recommended, especially if they can be triggered by regular users, since a single user can block your entire application for the duration of the db export (and the export time will increase with the growth of your db).

I'm not sure what to recommend because I don't understand what actually is your use case, what your app does and how the user is intended to interact with the exported database (what if the collections have files?).

If you want each user to have their own database, then I would recommend to:
  - have separate PocketBase processes each with individual `pb_data` for each user
  - built your app from scratch with something else to accommodate better your specific use case

**Reply by ganigeorgiev** [reply]
Some security notes if you still want to proceed with the above code:
- Make sure that `userId` value can't be set to a random user by anyone (except maybe if the above action is executed by admins/superusers).
- You need to keep some sort of flag for the pending backups to prevent running multiple exports for the same user since they will overwrite the same db file.
- You may need to implement some sort of rate limit for the action (v0.23 has a builtin rate limiter) or maybe better - schedule an export job in the background and once completed send an email with protected link to the targeted user.
- Schedule a cleanup job to prevent old exports filling your disk space.

But again, you might be better with custom solution or multiple PocketBase `pb_data` per user as this will also cover uploaded files and it is less likely to affect (and accidentally export) other users data.

**Reply by lumosminima** [reply]
I was previously using dbx builder directly but the main reason for dao was the creation of db tables using ImportCollection() like following:
```go
package export

import (
	"encoding/json"
	"log"

	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase/daos"
	"github.com/pocketbase/pocketbase/models"
)

func NewScaffoldedDb(db dbx.Builder) error {
	jsonData := `[
			...
		]`

	collections := []*models.Collection{}
	if err := json.Unmarshal([]byte(jsonData), &collections); err != nil {
		return nil, err
	}

	newDb := daos.New(db)
	importErr := newDb.ImportCollections(collections, true, nil)
	if importErr != nil {
		return nil, importErr
	}
	return newDb, nil
}
```
Other option I'm currently testing is: `sqlite3 data.db .schema > data.sql` and using the sql after cleaning system tables. I have two usecases for this allowing users to export their data and local offline usage, in either cases the resulting db will be either be used by sqlite-wasm in browser using OPFS or in some other app for data analysis purposes. It is not used on server just generated and server remains source of truth.

You mentioned user blocking app while exporting but why would that be? I thought sqlite can have many concurrent readers and only writes lock and in case WAL mode is enabled then both can proceed concurrently no? Does copying data from attached db locks the attached db too? I have thought about using some other solution but wanted to first exhaust my options with pocketbase since it seemed like a simple enough usecase.

**Reply by lumosminima** [reply]
> Some security notes if you still want to proceed with the above code:
> 
>     * Make sure that `userId` value can't be set to a random user by anyone (except maybe if the above action is executed by admins/superusers).
> 
>     * You need to keep some sort of flag for the pending backups to prevent running multiple exports for the same user since they will overwrite the same db file.
> 
>     * You may need to implement some sort of rate limit for the action (v0.23 has a builtin rate limiter) or maybe better - schedule an export job in the background and once completed send an email with protected link to the targeted user.
> 
>     * Schedule a cleanup job to prevent old exports filling your disk space.
> 
> 
> But again, you might be better with custom solution or multiple PocketBase `pb_data` per user as this will also cover uploaded files and it is less likely to affect (and accidentally export) other users data.

This was an experimental handler to just see data flow working first without any additional processing, I would be doing proper access and CanAccessRecord checks later on if this ends up working out for me. Also, the each db file is removed first so it is always the freshest copy. The reason I'm not using seperate app instances is because that would be hard to manage since in my data model spaces can have multiple user and I would need to do a lot of glue work between different users that are part of some space unless I'm mistaken and this is not that complicated. In any case each users generated data will only be what they have access to based on their space_role and even before that I can just query the space collection to see if userId has some role and enforce auth.

**Reply by ganigeorgiev** [reply, marked as answer]
>  I thought sqlite can have many concurrent readers and only writes lock and in case WAL mode is enabled then both can proceed concurrently no?

No, because you have **mixed (read&write)  transaction** and this can throw `SQLITE_LOCKED` or `SQLITE_BUSY` depending on the other concurrent operations (PocketBase has auto/fail-retry in addition to the `busy_timeout` PRAGMA on app level but this only prevents surfacing the error and other users will still have to wait to execute their requests until the export is completed).
